### ISSUE TITLE
yum_package provides false positive for documented version matching syntax

### ISSUE THREAD

**Issue by jorhett** [chef/chef#3332]
From chef run, using the explicit syntax shown on https://docs.chef.io/resource_yum.html

Simplest possible test case:

```
$ cat test.rb
package 'php55 = 5.5.24-1.97.amzn1' do
    action :install
end
```

Running in zero mode for simplest test: 
    $ chef-client -z test.rb
    [2015-05-04T14:39:30-07:00] WARN: No config file found or specified on command line, using command line options.
    [2015-05-04T14:39:30-07:00] WARN: No cookbooks directory found at or above current directory. Assuming /home/jo.
    Starting Chef Client, version 11.18.6
    resolving cookbooks for run list: []
    Synchronizing Cookbooks:
    Compiling Cookbooks...
    [2015-05-04T14:39:31-07:00] WARN: Node **\* has an empty run list.
    Converging 1 resources
    Recipe: @recipe_files::/home/jo/test.rb
      \* package[php55 = 5.5.24-1.97.amzn1] action install (up to date)

```
Running handlers:
Running handlers complete
   Chef Client finished, 0/1 resources updated in 2.664947883 seconds
```

However this simply isn't true: 
    $ rpm -qa php55
    php55-5.5.21-1.96.amzn1.x86_64

```
$ yum list php55
Loaded plugins: priorities, s3iam, update-motd, upgrade-helper
1603 packages excluded due to repository priority protections
Installed Packages
php55.x86_64                               5.5.21-1.96.amzn1                                              @amzn-updates
Available Packages
php55.x86_64                               5.5.24-1.97.amzn1                                                 amzn-updates
```

This isn't a problem with zero mode, this was first observed in full client mode and reduced to this small recipe for smallest possible test.

**Comment by jorhett** [comment]
FYI debug output shows the failure quite clearly:

    [2015-05-04T15:42:16-07:00] INFO: Processing package[php55 = 5.5.24-1.97.amzn1] action install (@recipe_files::/home/jo/test.rb line 1)
    [2015-05-04T15:42:18-07:00] DEBUG: package[php55 = 5.5.24-1.97.amzn1] no package found for php55 = 5.5.24-1.97.amzn1 but matched Provides for php55
    [2015-05-04T15:42:18-07:00] DEBUG: package[php55 = 5.5.24-1.97.amzn1] checking yum info for php55
    [2015-05-04T15:42:18-07:00] DEBUG: package[php55 = 5.5.24-1.97.amzn1] installed version: 5.5.21-1.96.amzn1 candidate version: 5.5.24-1.97.amzn1
    [2015-05-04T15:42:18-07:00] DEBUG: package[php55 = 5.5.24-1.97.amzn1] is already installed - nothing to do

**Comment by jtimberman** [comment]
I did some more digging into this.

I have a Fedora 21 system that has wget installed, with a newer version available.

```
DEBUG: yum_package[wget] installed version: ["1.15-3.fc21"] candidate version: 1.16.3-1.fc21
```

The following syntaxes do NOT result in the newer version being installed:

```ruby
package "wget = 1.16.3"
package "wget >= 1.16.3"
package "wget" do
  version "1.16.3"
end

package "wget" do
  version "= 1.16.3"
end

package "wget = 1.16.3-1.fc21"
```

The only syntax that works is:

```ruby
package "wget" do
  version "1.16.3-1.fc21"
end
```

The yum provider is quite complex and is difficult to follow what's going on inside it. Because of how much it's trying to do, I couldn't quite determine why the "wget = VER" syntax didn't work. Here's some debug output from that though:

```
[vagrant@default-fedora-21 ~]$ sudo chef-apply -e 'package "wget = 1.16.3-1.fc21"'
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * yum_package[wget = 1.16.3-1.fc21] action install (up to date)
[vagrant@default-fedora-21 ~]$ rpm -q wget
wget-1.15-3.fc21.x86_64

Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * yum_package[wget = 1.16.3-1.fc21] action install[2015-05-04T23:59:48+00:00] INFO: Processing yum_package[wget = 1.16.3-1.fc21] action install ((chef-apply cookbook)::(chef-apply recipe) line 1)
[2015-05-04T23:59:48+00:00] DEBUG: providers for generic yum_package resource enabled on node include: [Chef::Provider::Package::Yum]
[2015-05-04T23:59:48+00:00] DEBUG: providers that refused resource yum_package[wget = 1.16.3-1.fc21] were: []
[2015-05-04T23:59:48+00:00] DEBUG: providers that support resource yum_package[wget = 1.16.3-1.fc21] include: [Chef::Provider::Package::Yum]
[2015-05-04T23:59:48+00:00] DEBUG: providers that survived replacement include: [Chef::Provider::Package::Yum]
.....omg this yum package list is huge
wget 0 1.15 3.fc21 x86_64 ['wget(x86-64) = 1.15-3.fc21', 'wget = 1.15-3.fc21', 'webclient', 'config(wget) = 1.15-3.fc21', 'bundled(gnulib)'] r fedora
wget 0 1.16.3 1.fc21 x86_64 [] a updates
[2015-05-04T23:59:55+00:00] DEBUG: wget = 1.16.3-1.fc21: Unable to match package 'wget = 1.16.3-1.fc21' but matched 1 , selected 'wget' version '1.16.3-1.fc21'
[2015-05-04T23:59:55+00:00] DEBUG: yum_package[wget = 1.16.3-1.fc21] checking yum info for wget-1.16.3-1.fc21
[2015-05-04T23:59:55+00:00] DEBUG: yum_package[wget = 1.16.3-1.fc21] installed version: ["1.15-3.fc21"] candidate version: 1.16.3-1.fc21
[2015-05-04T23:59:55+00:00] DEBUG: yum_package[wget = 1.16.3-1.fc21] wget 1.15-3.fc21 already installed
[2015-05-04T23:59:55+00:00] DEBUG: yum_package[wget = 1.16.3-1.fc21] is already installed - nothing to do
```

**Comment by jorhett** [comment]
I have the same situation with wget, and for more examples all of the following work in pure Yum:

Matching partial version:
    # yum install "wget == 1.16"
    Loaded plugins: priorities, s3iam, update-motd, upgrade-helper
    1603 packages excluded due to repository priority protections
    Resolving Dependencies
    --> Running transaction check
    ---> Package wget.x86_64 0:1.15-3.12.amzn1 will be updated
    ---> Package wget.x86_64 0:1.16-3.16.amzn1 will be an update

Matching minimum version:
    # yum install "wget >= 1.15"
    Loaded plugins: priorities, s3iam, update-motd, upgrade-helper
    1603 packages excluded due to repository priority protections
    Resolving Dependencies
    --> Running transaction check
    ---> Package wget.x86_64 0:1.15-3.12.amzn1 will be updated
    ---> Package wget.x86_64 0:1.16-3.16.amzn1 will be an update

So yum understands these versions at the system level just fine, but:
    # chef-apply --l debug -e 'package "wget = 1.16"' |grep wget
    [2015-05-04T17:12:23-07:00] INFO: Processing package[wget = 1.16] action install ((chef-apply cookbook)::(chef-apply recipe) line 1)
    [2015-05-04T17:12:25-07:00] DEBUG: package[wget = 1.16] no package found for wget = 1.16 but matched Provides for wget
    [2015-05-04T17:12:25-07:00] WARN: package[wget = 1.16] matched multiple Provides for wget = 1.16 but we can only use the first match: wget. Please use a more specific version.
    [2015-05-04T17:12:25-07:00] DEBUG: package[wget = 1.16] checking yum info for wget
    [2015-05-04T17:12:25-07:00] DEBUG: package[wget = 1.16] installed version: 1.15-3.12.amzn1 candidate version: 1.16-3.16.amzn1
    [2015-05-04T17:12:25-07:00] DEBUG: package[wget = 1.16] is already installed - nothing to do

**Comment by chilicheech** [comment]
Having a similar issue where when I run yum I get this:
```
$ sudo yum install "openssl >= 1.0.1e-16.el7_5.15.x86_64"
Resolving Dependencies
--> Running transaction check
---> Package openssl.x86_64 1:1.0.1e-42.el7_1.5 will be updated
---> Package openssl.x86_64 1:1.0.1e-42.el7_1.6 will be an update
--> Processing Dependency: openssl-libs(x86-64) = 1:1.0.1e-42.el7_1.6 for package: 1:openssl-1.0.1e-42.el7_1.6.x86_64
--> Running transaction check
---> Package openssl-libs.x86_64 1:1.0.1e-42.el7_1.5 will be updated
---> Package openssl-libs.x86_64 1:1.0.1e-42.el7_1.6 will be an update
--> Finished Dependency Resolution
```
And when I run the yum_package
```
$ chef-apply --l debug -e 'yum_package "openssl >= 1.0.1e-16.el7_5.15.x86_64"' | grep ope
[2015-07-13T18:27:22+00:00] INFO: Processing yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] action install ((chef-apply cookbook)::(chef-apply recipe) line 1)
[2015-07-13T18:27:22+00:00] DEBUG: Provider for action install on resource yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] is Chef::Provider::Package::Yum
[2015-07-13T18:27:23+00:00] DEBUG: openssl >= 1.0.1e-16.el7_5.15.x86_64: Unable to match package 'openssl >= 1.0.1e-16.el7_5.15.x86_64' but matched 9 packages, selected 'openssl' version '1.0.1e-34.el7_0.4'
[2015-07-13T18:27:23+00:00] WARN: yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] matched multiple Provides for openssl >= 1.0.1e-16.el7_5.15.x86_64 but we can only use the first match: openssl. Please use a more specific version.
[2015-07-13T18:27:23+00:00] DEBUG: yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] checking yum info for openssl-1.0.1e-34.el7_0.4
[2015-07-13T18:27:23+00:00] DEBUG: yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] installed version: ["1.0.1e-42.el7_1.5"] candidate version: 1.0.1e-42.el7_1.6
[2015-07-13T18:27:23+00:00] DEBUG: yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] openssl 1.0.1e-42.el7_1.5 needs updating to 1.0.1e-34.el7_0.4
Error executing action `install` on resource 'yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64]'
Installed package openssl-1.0.1e-42.el7_1.5 is newer than candidate package openssl-1.0.1e-34.el7_0.4
  1: yum_package "openssl >= 1.0.1e-16.el7_5.15.x86_64"
yum_package("openssl >= 1.0.1e-16.el7_5.15.x86_64") do
  package_name "openssl"
[2015-07-13T18:27:23+00:00] DEBUG: Chef::Exceptions::Package: yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] ((chef-apply cookbook)::(chef-apply recipe) line 1) had an error: Chef::Exceptions::Package: Installed package openssl-1.0.1e-42.el7_1.5 is newer than candidate package openssl-1.0.1e-34.el7_0.4
[2015-07-13T18:27:23+00:00] FATAL: Chef::Exceptions::Package: yum_package[openssl >= 1.0.1e-16.el7_5.15.x86_64] ((chef-apply cookbook)::(chef-apply recipe) line 1) had an error: Chef::Exceptions::Package: Installed package openssl-1.0.1e-42.el7_1.5 is newer than candidate package openssl-1.0.1e-34.el7_0.4
$
```
What I don't get is that the provider says it's selecting version 1.0.1e-34.el7_0.4 and then later it says the candidate version is 1.0.1e-42.el7_1.6 and at the end it says the candidate version is 1.0.1e-34.el7_0.4. What's going on with how it selects the version? Should I drop yum_package in favor of execute 'yum install "openssl >= 1.0.1e-16.el7_5.15.x86_64"'?

**Comment by pburkholder** [comment]
I had user with the same issue with `httpd`.  

The problem seems to be rooted at  https://github.com/chef/chef/blob/master/lib/chef/provider/package/yum.rb#L1391, where the first package returned from `packages` is deemed to be the correct one.

The list of `packages` is pulled from a call out `yum-dump.py`.  When the yumdump results are matched against `httpd > 2.2.15-47.el6_7` you'll get, based on the `yum-dump` results in this case:,

```
httpd 0 2.2.15 47.el6_7.2 x86_64 [] a sat-mirror 
httpd 0 2.2.15 47.el6_7.1 x86_64 [] a sat-mirror 
httpd 0 2.2.15 47.el6_7.3 x86_64 [] r sat-mirror 
```

So the first, release 47.el6_7.2, is selected as the candidate, but it's older than the installed version so.. fail.

**Comment by lamont-granquist** [comment]
Its probably better to use the `:upgrade` action when it makes sense here.

The `:install` action is fundamentally written around the `if its already installed, don't do anything`, and the `:upgrade` action is written around the "install the best package" use case.  Any package installation with an equality version pin will actually work better with both `allow_downgrade true` and the version pin in the `version` property.

**Comment by lamont-granquist** [comment]
looks like i can actually fix this in #6540
